extras/python/audio2mozzi.py

```diff
@@ -0,0 +1,167 @@
+(beneath the editing window) to samples. Then you can generate
+and save 1 second of a waveform and it will fit your table
+length.
+
+To convert samples using Audacity:
+For a recorded audio sample, set the project rate to the
+MOZZI_AUDIO_RATE (16384 in the current version).
+Samples can be any length, as long as they fit in your Arduino.
+Save by exporting with the format set to "Other uncompressed formats",
+"Header: RAW(headerless)" and choose the encoding you prefer (Signed 8/16/32-bit PCM or 32-bit Float).
+
+Alternative to convert samples via CLI using sox (http://sox.sourceforge.net/):
+sox <inputfile> -b <8/16/32> -c 1 -e <floating-point/signed-integer> -r 16384 <outputfile>
+
+Now use the file you just exported, as the "input_file" to convert and
+set the other parameters according to what you chose for conversion.
+"""
+
+import array
+import math
+import random
+import sys
+import textwrap
+from argparse import ArgumentParser
+from pathlib import Path
+
+
+def float2mozzi(args):
+    input_path = args.input_file.expanduser()
+    output_path = (
+        args.output_file.expanduser()
+        if args.output_file is not None
+        else input_path.with_suffix(".h")
+    )
+
+    with input_path.open("rb") as fin, output_path.open("w") as fout:
+        print(f"opened {input_path}")
+        num_input_values = int(
+            input_path.stat().st_size / (args.input_bits / 8)
+        )  # Adjust for number format (table at top of https://docs.python.org/3/library/array.html)
+
+        array_type = ""
+        if args.input_bits == 8:
+            array_type = "b"
+        elif args.input_bits == 16:
+            array_type = "h"
+        elif args.input_bits == 32:
+            array_type = "f" if args.input_encoding == "float" else "i"
+
+        valuesfromfile = array.array(array_type)
+        try:
+            valuesfromfile.fromfile(fin, num_input_values)
+        except EOFError:
+            pass
+        in_values = valuesfromfile.tolist()
+
+        tablename = (
+            args.table_name
+            if args.table_name is not None
+            else output_path.stem.replace("-", "_").upper()
+        )
+
+        fout.write(f"#ifndef {tablename}_H_\n")
+        fout.write(f"#define {tablename}_H_\n\n")
+        fout.write("#include <Arduino.h>\n")
+        fout.write('#include "mozzi_pgmspace.h"\n\n')
+        fout.write(f"#define {tablename}_NUM_CELLS {len(in_values)}\n")
+        fout.write(f"#define {tablename}_SAMPLERATE {args.sample_rate}\n\n")
+
+        table = f"CONSTTABLE_STORAGE(int{args.output_bits}_t) {tablename}_DATA [] = {{"
+        max_output_value = 2 << (args.output_bits - 2)  # Halved because signed
+        out_values = []
+        for value in in_values:
+            cnt_33 = 0
+            out_values.append(
+                math.trunc((value * max_output_value) + 0.5)
+                if args.input_encoding == "float"
+                else value
```

Comment:
Shouldn't some kind of normalization happen here in case of 16bits as input? From my tests, inputting a 16bits and expecting an 8bits output produces some way out of range values.

Reply:
Yeah, you are right, I didn't test it properly. Using some help from Numpy 😄